plugins/X3DReader/X3DReader.py

```diff
@@ -0,0 +1,902 @@
+            self.transform.setByScaleFactor(scale)
+            self.index_base = 0
+            
+            # Traverse the scene tree, populate the shapes list
+            self.processChildNodes(xml_scene)
+            
+            if self.shapes:
+                bui = MeshBuilder()
+                bui.setVertices(numpy.concatenate([shape.verts for shape in self.shapes]))
+                bui.setIndices(numpy.concatenate([shape.faces for shape in self.shapes]))
+                bui.calculateNormals()
+                bui.setFileName(file_name)
+                    
+                scene = SceneNode()
+                scene.setMeshData(bui.build().getTransformed(Matrix()))
+                scene.setSelectable(True)
+                scene.setName(file_name)
+            else:
+                return None
+            
+        except Exception as e:
+            Logger.log("e", "exception occured in x3d reader: %s", e)
+
+        try:
+            boundingBox = scene.getBoundingBox()
+            boundingBox.isValid()
+        except:
+            return None
+
+        return scene
+    
+    # ------------------------- XML tree traversal
+  
+    def processNode(self, xml_node):
+        xml_node =  self.resolveDefUse(xml_node)
+        if xml_node is None:
+            return
+        
+        tag = xml_node.tag
+        if tag in ("Group", "StaticGroup", "CADAssembly", "CADFace", "CADLayer", "Collision"):
+            self.processChildNodes(xml_node)
+        if tag == "CADPart":
+            self.processTransform(xml_node) # TODO: split the parts
+        elif tag == "LOD":
+            self.processNode(xml_node[0])
+        elif tag == "Transform":
+            self.processTransform(xml_node)
+        elif tag == "Shape":
+            self.processShape(xml_node)
+            
+            
+    def processShape(self, xml_node):
+        # Find the geometry and the appearance inside the Shape
+        geometry = appearance = None
+        for subNode in xml_node:
+            if subNode.tag == "Appearance" and not appearance:
+                appearance = self.resolveDefUse(subNode)
+            elif subNode.tag in self.geometry_importers and not geometry:
+                geometry = self.resolveDefUse(subNode)
+        
+        # TODO: appearance is completely ignored. At least apply the material color...        
+        if not geometry is None:
+            try:
+                self.verts = self.faces = [] # Safeguard 
+                self.geometry_importers[geometry.tag](self, geometry)
+                m = self.transform.getData()
+                verts = numpy.array([m.dot(vert)[:3] for vert in self.verts])
+                self.shapes.append(Shape(verts, self.faces, self.index_base, geometry.tag))
+                self.index_base += len(verts)
+                
+            except Exception as e:
+                Logger.log("e", "exception occured in x3d reader while reading %s: %s", geometry.tag, e)
```

Comment:
Have a look at Logger.logException. That logging already handles the printing of exception so you don't have to.